File: Iterator.py
# import collections.abc

# class ConcreteAggregate(collections.abc.Iterable):
#   """
#   Implement the Iterator creation interface to return an instance of the proper ConcreteIterator
#   """

#   def __init__(self):
#     self._data = None

#   def __iter__(self):
#     return ConcreteIterator(self)

# class ConcreteIterator(collections.abc.Iterator):
#   """
#   Implement the Iterator interface
#   """

#   def __init__(self, concrete_aggregate):
#     self._concrete_aggregate = concrete_aggregate

#   def __next__(self):
#     if True: # if no_elements_to_traverse:
#       raise StopIteration
#     return None # return element


# concrete_aggregate = ConcreteAggregate()
# for _ in concrete_aggregate:
#   print(_)

import random

class DragonArmy:
  def __init__(self):
    self._sizes = ['big', 'medium', 'small']
    self._types = ['water', 'earth', 'fire', 'air']

  def __iter__(self):
    return self

  def __next__(self):
    if random.randint(1,101) > 75:
      raise StopIteration
    
    return self._makeDragon()

  def _makeDragon(self):
    size = self._sizes[random.randint(0, 2)]
    type = self._types[random.randint(0,3)]

    return ' '.join([size, type, 'dragon'])

dragonArmy = DragonArmy()

dragons = []

for dragon in dragonArmy:
  print(dragon)

File: Iterator.js
// const dragons = [
//   'dragon 1',
//   'dragon 2',
//   'dragon 3'
// ];

// const iterator = dragons[Symbol.iterator]();
// iterator.next(); // {value: 'dragon 1', done: false}
// iterator.next(); // {value: 'dragon 2', done: false}
// iterator.next(); // {value: 'dragon 3', done: false}
// iterator.next(); // {value: undefined, done: true}

// // For...of uses it under the hood
// for (const dragon of dragons) {
//   dragon 
// };

// for (const char of dragons[0]) {
//   char 
// };

function randomItem(array) {

  const randomNumber = (min, max) => Math.floor(Math.random() * max) + min;
  const randomIndex = randomNumber(0, array.length);

  return array[randomIndex];
}

const makeDragon = () => {
  const dragonSizes = ['big', 'medium', 'small'];
  const dragonTypes = ['earth', 'wind', 'fire'];
  return `${randomItem(dragonSizes)} ${randomItem(dragonTypes)} dragon`;
}

const dragonArmy = {
  [Symbol.iterator]: () => {
    return {
      next: () => {
        const enoughDragonsSpawned = Math.random() > 0.75;
        if(!enoughDragonsSpawned)
          return {
            value: makeDragon(),
            done: false
          };
        return { done: true };
      }
    }
  }
}

for (const dragon of dragonArmy) {
  console.log(dragon);
}
